File: app/src/main/java/com/matchthree/game/engine/SpecialRules.kt
package com.matchthree.game.engine

import com.matchthree.game.model.Board
import com.matchthree.game.model.Gem
import com.matchthree.game.model.Position
import com.matchthree.game.model.Special

/**
 * Pure M4 special-gem rules (MECHANICS.md). All functions are deterministic and
 * JVM-testable — the engine calls these, the tests call these directly.
 *
 * Birth is per shape (MECHANICS.md): runs sharing cells form one shape, and
 * each shape births exactly one special by precedence **5-in-row > T/L >
 * 4-in-row > plain 3** — one matched gem transforms, the rest clear normally.
 * Non-overlapping shapes resolve independently, so one cascade round can birth
 * several specials.
 */
object SpecialRules {

    /** A gem chosen to transform into a special during a cascade round. */
    data class Birth(val special: Special, val gemId: Int, val cell: Position)

    /** True when the swap of [gemA] and [gemB] is legal even without a match. */
    fun swapContactLegal(gemA: Gem?, gemB: Gem?): Boolean {
        val specA = gemA?.special
        val specB = gemB?.special
        return (specA != null && specB != null) || specA == Special.HYPERCUBE || specB == Special.HYPERCUBE
    }

    /**
     * Decides the specials born from one cascade round: runs are clustered into
     * shapes by shared cells, and each shape births at most one special by
     * precedence (5-run > T/L > 4-run; plain 3-runs birth nothing). Groups are
     * ordered by their first run's index in [matches] — deterministic.
     */
    fun resolveBirths(board: Board, matches: List<Match>): List<Birth> =
        shapeGroups(matches).mapNotNull { birthForShape(board, it) }

    /**
     * Clusters runs into shapes: runs sharing any cell belong to the same shape
     * (a T/L is one shape of two intersecting runs; a shared cell implies the
     * same gem and color, since a cell holds one gem).
     */
    private fun shapeGroups(matches: List<Match>): List<List<Match>> {
        val parent = IntArray(matches.size) { it }
        fun find(i: Int): Int {
            var root = i
            while (parent[root] != root) root = parent[root]
            var cur = i
            while (parent[cur] != root) {
                val next = parent[cur]
                parent[cur] = root
                cur = next
            }
            return root
        }

        val cellOwner = HashMap<Position, Int>()
        for ((index, match) in matches.withIndex()) {
            for (pos in match.positions) {
                val owner = cellOwner.put(pos, index)
                if (owner != null) {
                    val a = find(owner)
                    val b = find(index)
                    if (a != b) parent[maxOf(a, b)] = minOf(a, b)
                }
            }
        }

        val groups = LinkedHashMap<Int, MutableList<Match>>()
        for ((index, match) in matches.withIndex()) {
            groups.getOrPut(find(index)) { mutableListOf() }.add(match)
        }
        return groups.values.toList()
    }

    /** The one special born from a single shape (or null for plain 3-runs). */
    private fun birthForShape(board: Board, runs: List<Match>): Birth? {
        // 1. 5-in-row wins over everything (including a crossing T/L).
        val five = runs.firstOrNull { it.positions.size >= 5 }
        if (five != null) {
            val cell = five.positions.firstOrNull { board.gemAt(it)?.special == null }
                ?: five.positions[2]
            val gem = checkNotNull(board.gemAt(cell)) { "matched run cell $cell is empty" }
            return Birth(Special.HYPERCUBE, gem.id, cell)
        }

        // 2. T/L: a position shared by a horizontal and a vertical run.
        val cross = intersectionCell(board, runs)
        if (cross != null) {
            val gem = checkNotNull(board.gemAt(cross)) { "intersection cell $cross is empty" }
            return Birth(Special.STAR, gem.id, cross)
        }

        // 3. 4-in-row becomes a Flame.
        val four = runs.firstOrNull { it.positions.size == 4 }
        if (four != null) {
            val cell = four.positions.firstOrNull { board.gemAt(it)?.special == null } ?: four.positions[1]
            val gem = checkNotNull(board.gemAt(cell)) { "matched run cell $cell is empty" }
            return Birth(Special.FLAME, gem.id, cell)
        }

        // 4. Plain 3-runs birth nothing.
        return null
    }

    /**
     * Cells cleared when [special] detonates around [center]: Flame = 3x3,
     * Star = its row + column. Hypercubes never reach this path (colorless:
     * they cannot be part of a match), so they add nothing here.
     */
    fun detonationCells(board: Board, center: Position, special: Special): Set<Position> =
        when (special) {
            Special.FLAME -> blast(center, radius = 1, board)
            Special.STAR -> starLines(center, board)
            Special.HYPERCUBE -> emptySet()
        }

    /**
     * Extra detonations from specials swept into a cascade round's match cells.
     * The matched cells themselves are cleared anyway; this adds their blasts.
     */
    fun sweptBlastCells(board: Board, matched: Set<Position>): Set<Position> {
        val extra = mutableSetOf<Position>()
        for (pos in matched) {
            val gem = board.gemAt(pos) ?: continue
            if (gem.special != null && gem.special != Special.HYPERCUBE) {
                extra += detonationCells(board, pos, gem.special)
            }
        }
        return extra
    }

    /**
     * Cells cleared when [hyperPos]'s hypercube triggers against the gem at
     * [partnerPos] (MECHANICS.md trigger rule and combo table). A plain partner
     * clears every gem of its color; a special partner powers every gem of its
     * color up into that special, which then detonates. The hypercube itself is
     * always consumed and included.
     */
    fun hypercubeTriggerCells(
        board: Board,
        hyperPos: Position,
        partnerPos: Position,
        partnerSpecial: Special?,
    ): Set<Position> {
        val partnerGem = board.gemAt(partnerPos)
            ?: return setOf(hyperPos)
        val color = partnerGem.type
        val affected = mutableSetOf(hyperPos)
        for (pos in board.positions()) {
            if (board.gemAt(pos)?.type != color) continue
            if (partnerSpecial == null) affected += pos
            else affected += detonationCells(board, pos, partnerSpecial)
        }
        return affected
    }

    /**
     * The cells cleared by a player-swap combo (MECHANICS.md combo table).
     * Both swapped positions count as cleared (the specials are consumed).
     */
    fun comboAffectedCells(
        board: Board,
        swapA: Position,
        swapB: Position,
        specA: Special,
        specB: Special,
    ): Set<Position> {
        // Hypercube combos: every gem of the swapped partner's color powers up
        // into the partner's special, then detonates simultaneously; the
        // hypercube itself is consumed too.
        if (specA == Special.HYPERCUBE || specB == Special.HYPERCUBE) {
            if (specA == Special.HYPERCUBE && specB == Special.HYPERCUBE) {
                return board.positions().toSet()
            }
            val hyperPos = if (specA == Special.HYPERCUBE) swapA else swapB
            val partnerPos = if (hyperPos == swapA) swapB else swapA
            val partnerSpec = if (specA == Special.HYPERCUBE) specB else specA
            return hypercubeTriggerCells(board, hyperPos, partnerPos, partnerSpec)
        }

        return when (setOf(specA, specB)) {
            setOf(Special.FLAME, Special.FLAME) -> blast(swapA, radius = 2, board)
            setOf(Special.FLAME, Special.STAR) -> thickCross(swapA, board)
            setOf(Special.STAR, Special.STAR) ->
                starLines(swapA, board) + starLines(swapB, board)
            else -> emptySet()
        }
    }

    // --- shape analysis -----------------------------------------------------

    private fun intersectionCell(board: Board, matches: List<Match>): Position? {
        val horizontal = matches.filter { it.positions.all { p -> p.row == it.positions.first().row } }
        val vertical = matches.filter { it.positions.all { p -> p.col == it.positions.first().col } }
        for (h in horizontal) {
            for (v in vertical) {
                for (pos in h.positions) {
                    if (pos in v.positions && board.gemAt(pos)?.special == null) return pos
                }
            }
        }
        return null
    }

    // --- area helpers -------------------------------------------------------

    private fun blast(center: Position, radius: Int, board: Board): Set<Position> {
        val cells = mutableSetOf<Position>()
        for (dr in -radius..radius) {
            for (dc in -radius..radius) {
                val pos = Position(center.row + dr, center.col + dc)
                if (board.isInside(pos)) cells += pos
            }
        }
        return cells
    }

    private fun starLines(center: Position, board: Board): Set<Position> {
        val cells = mutableSetOf<Position>()
        for (col in 0 until board.width) cells += Position(center.row, col)
        for (row in 0 until board.height) cells += Position(row, center.col)
        return cells
    }

    /** The 3-wide row + 3-wide column "thick cross" through [center]. */
    private fun thickCross(center: Position, board: Board): Set<Position> {
        val cells = mutableSetOf<Position>()
        for (dc in -1..1) {
            for (col in 0 until board.width) {
                val pos = Position(center.row + dc, col)
                if (board.isInside(pos)) cells += pos
            }
        }
        for (dr in -1..1) {
            for (row in 0 until board.height) {
                val pos = Position(row, center.col + dr)
                if (board.isInside(pos)) cells += pos
            }
        }
        return cells
    }
}

File: Engine/Rules/SpecialRules.cs
using MatchThree.Engine.Model;

namespace MatchThree.Engine.Rules;

public static class SpecialRules
{
    public sealed record Birth(Special Special, int GemId, Position Cell);

    public static bool SwapContactLegal(Gem? gemA, Gem? gemB)
    {
        var specA = gemA?.Special;
        var specB = gemB?.Special;
        return (specA is not null && specB is not null) ||
               specA == Special.Hypercube ||
               specB == Special.Hypercube;
    }

    public static List<Birth> ResolveBirths(Board board, List<Match> matches) =>
        ShapeGroups(matches)
            .Select(group => BirthForShape(board, group))
            .Where(birth => birth is not null)
            .Select(birth => birth!)
            .ToList();

    private static List<List<Match>> ShapeGroups(List<Match> matches)
    {
        var parent = new int[matches.Count];
        for (var i = 0; i < parent.Length; i++) parent[i] = i;

        int Find(int i)
        {
            var root = i;
            while (parent[root] != root) root = parent[root];
            var cur = i;
            while (parent[cur] != root)
            {
                var next = parent[cur];
                parent[cur] = root;
                cur = next;
            }
            return root;
        }

        var cellOwner = new Dictionary<Position, int>();
        for (var index = 0; index < matches.Count; index++)
        {
            foreach (var pos in matches[index].Positions)
            {
                if (cellOwner.TryGetValue(pos, out var owner))
                {
                    var a = Find(owner);
                    var b = Find(index);
                    if (a != b) parent[Math.Max(a, b)] = Math.Min(a, b);
                }
                else
                {
                    cellOwner[pos] = index;
                }
            }
        }

        var groups = new List<List<Match>>();
        var groupByRoot = new Dictionary<int, int>();
        for (var index = 0; index < matches.Count; index++)
        {
            var root = Find(index);
            if (!groupByRoot.TryGetValue(root, out var groupIndex))
            {
                groupIndex = groups.Count;
                groupByRoot[root] = groupIndex;
                groups.Add(new List<Match>());
            }
            groups[groupIndex].Add(matches[index]);
        }
        return groups;
    }

    private static Birth? BirthForShape(Board board, List<Match> runs)
    {
        var five = runs.FirstOrDefault(m => m.Positions.Count >= 5);
        if (five is not null)
        {
            var cell = FindFirstPlainCell(five.Positions, board) ?? five.Positions[2];
            var gem = board.GemAt(cell)
                ?? throw new InvalidOperationException($"matched run cell {cell} is empty");
            return new Birth(Special.Hypercube, gem.Id, cell);
        }

        var cross = IntersectionCell(board, runs);
        if (cross is not null)
        {
            var gem = board.GemAt(cross.Value)
                ?? throw new InvalidOperationException($"intersection cell {cross} is empty");
            return new Birth(Special.Star, gem.Id, cross.Value);
        }

        var four = runs.FirstOrDefault(m => m.Positions.Count == 4);
        if (four is not null)
        {
            var cell = FindFirstPlainCell(four.Positions, board) ?? four.Positions[1];
            var gem = board.GemAt(cell)
                ?? throw new InvalidOperationException($"matched run cell {cell} is empty");
            return new Birth(Special.Flame, gem.Id, cell);
        }

        return null;
    }

    public static ISet<Position> DetonationCells(Board board, Position center, Special special) =>
        special switch
        {
            Special.Flame => Blast(center, radius: 1, board),
            Special.Star => StarLines(center, board),
            Special.Hypercube => new HashSet<Position>(),
            _ => throw new ArgumentOutOfRangeException(nameof(special)),
        };

    public static ISet<Position> SweptBlastCells(Board board, ISet<Position> matched) =>
        ChainReactionCells(
            board,
            initiallyCleared: matched,
            initialDetonations: Array.Empty<DetonationSeed>())
        .Except(matched)
        .ToHashSet();

    public static ISet<Position> HypercubeTriggerCells(
        Board board,
        Position hyperPos,
        Position partnerPos,
        Special? partnerSpecial)
    {
        var partnerGem = board.GemAt(partnerPos);
        if (partnerGem is null) return new HashSet<Position> { hyperPos };

        var color = partnerGem.Value.Type;
        if (partnerSpecial is null) return HypercubeTriggerCellsByColor(board, hyperPos, color, null);

        var transformed = board.Positions()
            .Where(pos =>
                pos != hyperPos &&
                board.GemAt(pos) is { Type: var type, Special: not Special.Hypercube } &&
                type == color)
            .ToHashSet();

        return ChainReactionCells(
            board,
            initiallyCleared: new[] { hyperPos },
            initialDetonations: transformed.Select(pos => new DetonationSeed(pos, partnerSpecial.Value, color)),
            ignoredDetonators: transformed);
    }

    public static ISet<Position> ComboAffectedCells(
        Board board,
        Position swapA,
        Position swapB,
        Special specA,
        Special specB)
    {
        if (specA == Special.Hypercube || specB == Special.Hypercube)
        {
            if (specA == Special.Hypercube && specB == Special.Hypercube)
            {
                return board.Positions().ToHashSet();
            }
            var hyperPos = specA == Special.Hypercube ? swapA : swapB;
            var partnerPos = hyperPos == swapA ? swapB : swapA;
            var partnerSpec = specA == Special.Hypercube ? specB : specA;
            return HypercubeTriggerCells(board, hyperPos, partnerPos, partnerSpec);
        }

        var affected = (specA, specB) switch
        {
            (Special.Flame, Special.Flame) => Blast(swapA, radius: 2, board),
            (Special.Flame, Special.Star) or (Special.Star, Special.Flame) => ThickCross(swapA, board),
            (Special.Star, Special.Star) => StarLines(swapA, board).Union(StarLines(swapB, board)).ToHashSet(),
            _ => new HashSet<Position>(),
        };

        return ChainReactionCells(
            board,
            initiallyCleared: affected,
            initialDetonations: Array.Empty<DetonationSeed>(),
            ignoredDetonators: new HashSet<Position> { swapA, swapB });
    }

    private sealed record DetonationSeed(Position Center, Special Effect, GemType Color);

    private static ISet<Position> ChainReactionCells(
        Board board,
        IEnumerable<Position> initiallyCleared,
        IEnumerable<DetonationSeed> initialDetonations,
        ISet<Position>? ignoredDetonators = null)
    {
        var ignored = ignoredDetonators ?? new HashSet<Position>();
        var affected = initiallyCleared.ToHashSet();
        var pending = new Queue<DetonationSeed>();
        var queuedDetonators = new HashSet<Position>();
        var triggeredHypercubes = new HashSet<Position>();

        void EnqueueSeed(DetonationSeed seed)
        {
            if (queuedDetonators.Add(seed.Center)) pending.Enqueue(seed);
        }

        void EnqueueActualDetonator(Position pos)
        {
            if (ignored.Contains(pos)) return;
            var gem = board.GemAt(pos);
            if (gem is { Special: { } special, Type: var color } && special != Special.Hypercube)
            {
                EnqueueSeed(new DetonationSeed(pos, special, color));
            }
        }

        foreach (var seed in initialDetonations.OrderBy(seed => seed.Center.Row).ThenBy(seed => seed.Center.Col))
        {
            EnqueueSeed(seed);
        }
        foreach (var pos in OrderByBoard(initiallyCleared))
        {
            EnqueueActualDetonator(pos);
        }

        while (pending.Count > 0)
        {
            var seed = pending.Dequeue();
            foreach (var cell in OrderByBoard(DetonationCells(board, seed.Center, seed.Effect)))
            {
                affected.Add(cell);

                var gem = board.GemAt(cell);
                if (gem is not { Special: { } special }) continue;

                if (special == Special.Hypercube)
                {
                    if (triggeredHypercubes.Add(cell))
                    {
                        affected.UnionWith(HypercubeTriggerCellsByColor(board, cell, seed.Color, null));
                    }
                }
                else
                {
                    EnqueueActualDetonator(cell);
                }
            }
        }

        return affected;
    }

    private static ISet<Position> HypercubeTriggerCellsByColor(
        Board board,
        Position hyperPos,
        GemType color,
        Special? partnerSpecial)
    {
        var affected = new HashSet<Position> { hyperPos };
        foreach (var pos in board.Positions())
        {
            var gem = board.GemAt(pos);
            if (gem is not { Type: var type, Special: not Special.Hypercube } || type != color) continue;
            if (partnerSpecial is null) affected.Add(pos);
            else affected.UnionWith(DetonationCells(board, pos, partnerSpecial.Value));
        }
        return affected;
    }

    private static IEnumerable<Position> OrderByBoard(IEnumerable<Position> cells) =>
        cells.OrderBy(pos => pos.Row).ThenBy(pos => pos.Col);

    private static Position? FindFirstPlainCell(List<Position> cells, Board board)
    {
        foreach (var pos in cells)
        {
            if (board.GemAt(pos)?.Special is null) return pos;
        }
        return null;
    }

    private static Position? IntersectionCell(Board board, List<Match> matches)
    {
        var horizontal = matches
            .Where(m => m.Positions.All(p => p.Row == m.Positions[0].Row))
            .ToList();
        var vertical = matches
            .Where(m => m.Positions.All(p => p.Col == m.Positions[0].Col))
            .ToList();

        foreach (var h in horizontal)
        {
            foreach (var v in vertical)
            {
                foreach (var pos in h.Positions)
                {
                    if (v.Positions.Contains(pos) && board.GemAt(pos)?.Special is null) return pos;
                }
            }
        }
        return null;
    }

    private static ISet<Position> Blast(Position center, int radius, Board board)
    {
        var cells = new HashSet<Position>();
        for (var dr = -radius; dr <= radius; dr++)
        {
            for (var dc = -radius; dc <= radius; dc++)
            {
                var pos = new Position(center.Row + dr, center.Col + dc);
                if (board.IsInside(pos)) cells.Add(pos);
            }
        }
        return cells;
    }

    private static ISet<Position> StarLines(Position center, Board board)
    {
        var cells = new HashSet<Position>();
        for (var col = 0; col < board.Width; col++) cells.Add(new Position(center.Row, col));
        for (var row = 0; row < board.Height; row++) cells.Add(new Position(row, center.Col));
        return cells;
    }

    private static ISet<Position> ThickCross(Position center, Board board)
    {
        var cells = new HashSet<Position>();
        for (var dc = -1; dc <= 1; dc++)
        {
            for (var col = 0; col < board.Width; col++)
            {
                var pos = new Position(center.Row + dc, col);
                if (board.IsInside(pos)) cells.Add(pos);
            }
        }
        for (var dr = -1; dr <= 1; dr++)
        {
            for (var row = 0; row < board.Height; row++)
            {
                var pos = new Position(row, center.Col + dr);
                if (board.IsInside(pos)) cells.Add(pos);
            }
        }
        return cells;
    }
}